File: app/src/androidTest/java/com/abishov/hexocat/common/rule/MockWebServerRule.java
package com.abishov.hexocat.common.rule;

import android.support.test.espresso.Espresso;
import android.support.test.espresso.IdlingResource;
import com.abishov.hexocat.HexocatTestApp;
import com.jakewharton.espresso.OkHttp3IdlingResource;
import io.appflate.restmock.RESTMockServer;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

public final class MockWebServerRule implements TestRule {

  @Override
  public Statement apply(Statement base, Description description) {
    return new Statement() {
      @Override
      public void evaluate() throws Throwable {
        HttpUrl baseUrl = HttpUrl.parse(RESTMockServer.getUrl());
        HexocatTestApp.overrideBaseUrl(baseUrl);

        RESTMockServer.reset();

        OkHttpClient client = HexocatTestApp.getInstance().appComponent().okHttpClient();
        IdlingResource okhttpIdlingResource = OkHttp3IdlingResource.create("okhttp", client);

        Espresso.registerIdlingResources(okhttpIdlingResource);
        try {
          base.evaluate();
        } finally {
          Espresso.unregisterIdlingResources(okhttpIdlingResource);
        }
      }
    };
  }
}

File: android/src/androidTest/java/com/abishov/hexocat/android/common/rule/MockWebServerRule.kt
package com.abishov.hexocat.android.common.rule

import androidx.test.espresso.IdlingRegistry
import com.abishov.hexocat.android.HexocatTestApp
import com.jakewharton.espresso.OkHttp3IdlingResource
import io.appflate.restmock.RESTMockServer
import okhttp3.HttpUrl.Companion.toHttpUrl
import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runners.model.Statement

class MockWebServerRule : TestRule {

    override fun apply(base: Statement, description: Description): Statement {
        return object : Statement() {

            @Throws(Throwable::class)
            override fun evaluate() {
                val baseUrl = RESTMockServer.getUrl().toHttpUrl()
                    .newBuilder()
                    .addPathSegment("graphql")
                    .build()

                HexocatTestApp.overrideBaseUrl(baseUrl)
                RESTMockServer.reset()

                val client = HexocatTestApp.instance.appComponent().okHttpClient()
                val okhttpIdlingResource = OkHttp3IdlingResource.create("okhttp", client)

                IdlingRegistry.getInstance().register(okhttpIdlingResource)
                try {
                    base.evaluate()
                } finally {
                    IdlingRegistry.getInstance().unregister(okhttpIdlingResource)
                }
            }
        }
    }
}
